modules/user_data/templates/07_cluster_setup.sh.tpl

```diff
@@ -0,0 +1,190 @@
+#!/usr/bin/env bash
+
+set -euo pipefail
+
+NODE_DNS=$(cat /tmp/node_dns)
+IMDS_TOKEN=$(curl -Ss -H "X-aws-ec2-metadata-token-ttl-seconds: 6000" -XPUT 169.254.169.254/latest/api/token)
+INSTANCE_ID=$(curl -Ss -H "X-aws-ec2-metadata-token: $IMDS_TOKEN" 169.254.169.254/latest/meta-data/instance-id)
+GRAPHDB_ADMIN_PASSWORD=$(aws --cli-connect-timeout 300 ssm get-parameter --region ${region} --name "/${name}/graphdb/admin_password" --with-decryption --query "Parameter.Value" --output text)
+VPC_ID=$(aws ec2 describe-instances --instance-id "$${INSTANCE_ID}" --query 'Reservations[0].Instances[0].VpcId' --output text)
+
+RETRY_DELAY=5
+MAX_RETRIES=10
+
+echo "###########################"
+echo "#    Starting GraphDB     #"
+echo "###########################"
+
+# the proxy service is set up in the AMI but not enabled, so we enable and start it
+systemctl daemon-reload
+systemctl start graphdb
+systemctl enable graphdb-cluster-proxy.service
+systemctl start graphdb-cluster-proxy.service
+
+#####################
+#   Cluster setup   #
+#####################
+
+# Checks if GraphDB is started, we assume it is when the infrastructure endpoint is reached
+check_gdb() {
+  if [ -z "$1" ]; then
+    echo "Error: IP address or hostname is not provided."
+    return 1
+  fi
+
+  local gdb_address="$1:7201/rest/monitor/infrastructure"
+  if curl -s --head -u "admin:$${GRAPHDB_ADMIN_PASSWORD}" --fail "$${gdb_address}" >/dev/null; then
+    echo "Success, GraphDB node $${gdb_address} is available"
+    return 0
+  else
+    echo "GraphDB node $${gdb_address} is not available yet"
+    return 1
+  fi
+}
+
+# Waits for 3 DNS records to be available
+wait_dns_records() {
+  ALL_DNS_RECORDS=($(aws ec2 describe-instances --filters "Name=vpc-id,Values=$${VPC_ID}" --query 'Reservations[*].Instances[*].[PrivateDnsName]' --output text))
+  ALL_DNS_RECORDS_COUNT="$${#ALL_DNS_RECORDS[@]}"
+
+  if [ "$${ALL_DNS_RECORDS_COUNT}" -ne 3 ]; then
+    sleep 5
+    wait_dns_records
+  else
+    echo "Private DNS zone record count is $${ALL_DNS_RECORDS_COUNT}"
+  fi
+}
+
+wait_dns_records
+
+# Check all instances are running
+for record in "$${ALL_DNS_RECORDS[@]}"; do
+  echo "Pinging $record"
+
+  if [ -n "$record" ]; then
+    while ! check_gdb "$record"; do
+      echo "Waiting for GDB $record to start"
+      sleep "$RETRY_DELAY"
+    done
+  else
+    echo "Error: address is empty."
+  fi
+done
+
+echo "All GDB instances are available."
+
+# Determine instance order
```

Comment:
Maybe one or two sentences how we chose what node creates the cluster would be useful wdyt?

Reply:
See line 82